### ISSUE TITLE
Weight update frequency of reward-modulated learning rule

### ISSUE THREAD

**Issue by godelicbach** [Hananel-Hazan/bindsnet#166]
```
for t in range(timesteps):
    for l in self.layers:
        # Update each layer of nodes.
        if isinstance(self.layers[l], AbstractInput):
            self.layers[l].forward(x=inpts[l][t])
        else:
            self.layers[l].forward(x=inpts[l])

        # Clamp neurons to spike.
        clamp = clamps.get(l, None)
        if clamp is not None:
            self.layers[l].s[clamp] = 1

    # Run synapse updates.
    for c in self.connections:
        self.connections[c].update(
            mask=masks.get(c, None), learning=self.learning, **kwargs
        )
```
Above is the part of the `run()` function in `network` class. I found that if I'm using MSTDPET learning rule, the network is updating the weight at every network-timestep with same reward value. I'm curious if it is intended or not because I expected the network to update at every agent-timestep (every time the agent interacts with the environment). 

After some thinking I see that updating weight at every network-timestep is required for non-reward modulated learning rules like STDP. So I think there should be some condition that enables or disables the weight updating. 

I'm thinking about adding new variable that checks whether the update is already done once or not. And this should be only activated for connections that follow reward-modulated learning rule.

**Comment by djsaunde** [comment]
This is intended; typically, the environment takes a step, and then the agent takes a step, then the environment, ... as in the typical RL setup. The network can "tick" (or, run a simulation timestep) multiple times when it's called in between calls to the environment.

It *might* be interesting to allow `reward` to be either a `float` or a `torch.Tensor`; that is, a sequence of rewards, one for every simulation timestep. This could encompass your use-case: set `reward = torch.tensor([1, 0, 0, 0, 0])` for a network with simulation time 5, and reward-modulated STDP with reward 1.

**Comment by djsaunde** [comment]
Hm, that's interesting. Could you put together a PR to this effect? I can imagine, like you say, that some learning rules could be applied in a batch-like manner, where the updates only occur after a chunk of simulation time (rather than per simulation timestep).

This is already possible in experimental scripts: you can perform arbitrary updates to connection weights of your `Network` object between calls to `pipeline.step()`. However, it would be nice to have some sort of functionality for this built into BindsNET.

**Comment by godelicbach** [comment]
OK,  I'll make PR when my implementation is stabilized enough. It may take some time.
